### ISSUE TITLE
jaq --version

### ISSUE THREAD

**Issue by pkoppstein** [01mf02/jaq#40]
`jaq --version` currently only gives the "release" version, so it is not possible to use the output of this flag to distinguish between various git commits.  

If it is too much trouble to give at least some part of the git sha, then it would still be helpful to distinguish the "official release" from a subsequent version.

**Comment by 01mf02** [comment]
There does not seem to be an easy way in the crate that provides the CLI (clap) to do this, AFAIK.
If you find a nice way, I would be happy to accept your PR.

**Comment by kklingenberg** [comment]
I think I might have hurried to close this when some nuance still remains. Consider this sequence (using, say, play-with-docker):

```bash
docker run --rm -it rust bash
apt remove -y git
cargo install --locked --git https://github.com/01mf02/jaq
```

which displays as its last line:

```text
   Installed package `jaq v1.0.0-alpha (https://github.com/01mf02/jaq#b072c621)` (executable `jaq`)
```

Which mentions the commit SHA. Sadly, then executing `jaq --version` predictably yields:

```text
jaq 1.0.0-alpha (rev: unknown)
```

I see now that depending on the git executable is subobptimal. I also noticed this crate: <https://docs.rs/built/latest/built/> which internally uses <https://docs.rs/git2/latest/git2/> to extract information from the repository. It might be best to rework my naive implementation into just using the `built` crate.

Also, I haven't been able to test the behaviour using homebrew. I hope it at least works when there's `git` in the PATH, but I wouldn't count on it.

**Comment by kklingenberg** [comment]
A followup to the above: I tried using the **git2** crate and reproduced the first scenario, and it works ok. But there's another complication: **git2** includes a -sys crate and it fails to build if the system doesn't include some toolchains.

So the failure mode is even harsher IMO, as jaq fails to build in systems where e.g. pkg-config isn't available. An easy-to-reproduce example is

```bash
docker run --rm rust:slim cargo install --locked --git https://github.com/kklingenberg/jaq --branch feature/use-git2
```

The tradeoffs never end! I'll refrain from sending another PR with overly complicated workarounds.

**Comment by jsejcksn** [comment]
Disclaimer: This is my first time engaging this project — I don't know its history or conventions. However, I have a suggestion:

> `jaq --version` currently only gives the "release" version, so it is not possible to use the output of this flag to distinguish between various git commits.

It seems to me the natural solution is to increment the version on every commit and build a release.

How this would be done is the part that requires more thinking. Here are my initial thoughts:

Only commits to the release target branch would qualify (I presume this is the [`main`](https://github.com/01mf02/jaq/tree/main) branch).

Incrementing of the version according to [semver](https://semver.org/) is possible to automate. By using [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) — which is _a specification for adding human and machine readable meaning to commit messages_ (or an equivalent system — the goal is simply consistently adhering to enumerations) — each commit message can be parsed to determine which version should be incremented (major/minor/patch). This could happen in a [git hook](https://git-scm.com/docs/githooks) (`pre-commit`?) or [GitHub Action](https://docs.github.com/en/actions).

**Comment by kklingenberg** [comment]
I agree @jsejcksn, using a pre-commit hook seems appropriate. I assumed that @pkoppstein wanted something similar to what gojq provides, as implemented in its [Makefile](https://github.com/itchyny/gojq/blob/d579009515e2d20a7227beb722cc784d6be15910/Makefile#L4). So I naively imitated it without considering the various installation scenarios.

**Comment by 01mf02** [comment]
Thanks @kklingenberg for reporting this.

As an immediate countermeasure, I reverted your original commit by bac33c3a1747ea390de368ebc64e9175b5218ce5.

I am skeptical towards an automatic increment of the version number on every commit. In particular because I estimate that doing this (correctly) would involve quite some work, which I would rather spend on improving jaq in other ways.

Would it help to have more frequent releases? My current efforts to decouple various parts of jaq should help to achieve this, and I plan to release a beta version once I have decided on proper crate names (#95) and once that the math filters have landed (#96).

**Comment by jsejcksn** [comment]
> I am skeptical towards an automatic increment of the version number on every commit. In particular because I estimate that doing this (correctly) would involve quite some work, which I would rather spend on improving jaq in other ways.

Of course — I understand this. As an intermediate step toward that longer-term goal, you could use the pre-commit hook to simply enforce that the source-committed version _does actually change_ when committing to the `main` branch — it will still be up to the committer to determine how to appropriately increment the version and manually do so prior to finalizing each commit.
